gl3n/aabb.d

```diff
@@ -1,204 +1,306 @@
-        assert(AABB(vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0), vec3(cast(at)1.0, cast(at)1.0, cast(at)1.0)).intersects(
-               AABB(vec3(cast(at)0.5, cast(at)0.5, cast(at)0.5), vec3(cast(at)3.0, cast(at)3.0, cast(at)3.0))));
+        alias AABB = AABBT!(at, dimension_);
+
+        assert(AABB(sizedVec([0.0, 0.0, 0.0, 0.0]), sizedVec([1.0, 1.0, 1.0, 1.0])).intersects(
+               AABB(sizedVec([0.5, 0.5, 0.5, 0.5]), sizedVec([3.0, 3.0, 3.0, 3.0]))));
 
-        assert(AABB(vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0), vec3(cast(at)1.0, cast(at)1.0, cast(at)1.0)).intersects(
-               AABB(vec3(cast(at)0.5, cast(at)0.5, cast(at)0.5), vec3(cast(at)1.7, cast(at)1.7, cast(at)1.7))));
+        assert(AABB(sizedVec([0.0, 0.0, 0.0, 0.0]), sizedVec([1.0, 1.0, 1.0, 1.0])).intersects(
+               AABB(sizedVec([0.5, 0.5, 0.5, 0.5]), sizedVec([1.7, 1.7, 1.7, 1.7]))));
 
-        assert(!AABB(vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0), vec3(cast(at)1.0, cast(at)1.0, cast(at)1.0)).intersects(
-                AABB(vec3(cast(at)2.5, cast(at)2.5, cast(at)2.5), vec3(cast(at)3.0, cast(at)3.0, cast(at)3.0))));
+        assert(!AABB(sizedVec([0.0, 0.0, 0.0, 0.0]), sizedVec([1.0, 1.0, 1.0, 1.0])).intersects(
+                AABB(sizedVec([2.5, 2.5, 2.5, 2.5]), sizedVec([3.0, 3.0, 3.0, 3.0]))));
     }
 
     /// Returns the extent of the AABB (also sometimes called size).
-    @property vec3 extent() const {
+    @property vec extent() const {
         return max - min;
     }
 
     /// Returns the half extent.
-    @property vec3 half_extent() const {
+    @property vec half_extent() const {
         return (max - min) / 2;
     }
 
     unittest {
-        alias AABBT!at AABB;
-        AABBT!at a = AABBT!at(vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0), vec3(cast(at)10.0, cast(at)10.0, cast(at)10.0));
-        assert(a.extent == vec3(cast(at)10.0, cast(at)10.0, cast(at)10.0));
+        alias AABB = AABBT!(at, dimension_);
+
+        AABB a = AABB(sizedVec([0.0, 0.0, 0.0, 0.0]), sizedVec([10.0, 10.0, 10.0, 10.0]));
+        assert(a.extent == sizedVec([10.0, 10.0, 10.0, 10.0]));
         assert(a.half_extent == a.extent / 2);
 
-        AABBT!at b = AABBT!at(vec3(cast(at)2.0, cast(at)2.0, cast(at)2.0), vec3(cast(at)10.0, cast(at)10.0, cast(at)10.0));
-        assert(b.extent == vec3(cast(at)8.0, cast(at)8.0, cast(at)8.0));
+        AABB b = AABB(sizedVec([2.0, 2.0, 2.0, 2.0]), sizedVec([10.0, 10.0, 10.0, 10.0]));
+        assert(b.extent == sizedVec([8.0, 8.0, 8.0, 8.0]));
         assert(b.half_extent == b.extent / 2);
-        
     }
 
     /// Returns the area of the AABB.
-    @property real area() const {
-        vec3 e = extent;
-        return 2.0 * (e.x * e.y + e.x * e.z + e.y * e.z);
-    }
-
-    unittest {
-        alias AABBT!at AABB;
-        AABB a = AABB(vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0), vec3(cast(at)1.0, cast(at)1.0, cast(at)1.0));
-        assert(a.area == 6);
+    static if (dimension_ <= 3) {
```

Comment:
I couldn't find a good way to calculate area of a N-dimension cube, so I restricted `area()` only for dimension 1, 2 and 3